### PR TITLE
Big http download upgrade#26717

### LINKED ISSUE #26717: Big HTTP Download - Upgrade
**Is your enhancement proposal related to a problem? Please describe.**

- [Big HTTP Download](https://docs.zephyrproject.org/latest/samples/net/sockets/big_http_download/README.html) example does not prints useful metrics (KiB, MiB, KiB/sec.) of download progress. It leads to time waste at development.

- [Big HTTP Download](https://docs.zephyrproject.org/latest/samples/net/sockets/big_http_download/README.html) example does not supports [Zephyr logging API](https://docs.zephyrproject.org/latest/reference/logging/index.html) properly. It leads to lost debug messages at dev time.

**Describe the solution you'd like**
The example project should become above upgrades.

### PR DESCRIPTION
**Pull request by KozhinovAlexander**
PR for Big HTTP Download - Upgrade 

The PR #26719 must be closed. All updates to code reviews were moved here also.

Fixes #26717

### CODE REVIEW

File: samples/net/sockets/big_http_download/src/big_http_download.c [hunk 1]
```diff
@@ -33,6 +33,11 @@
 
 #endif
 
+#define KiB	*1024
+#define MiB	*1024 KiB
+#define Bytes2KiB(Bytes)	(((float)Bytes)/(1.0f KiB))
+#define Bytes2MiB(Bytes)	(((float)Bytes)/(1.0f MiB))
+
```
Comment (Nukersson) [inline samples/net/sockets/big_http_download/src/big_http_download.c:44]: from @pfalcon in #26719 : I'm not sure about usage of floating-point arithmetic. This would be one cause of stack usage growth you faced. And it would grow binary size, and potentially make it not fit on smaller devices, which would be a loss of functionality.

Comment (Nukersson) [reply samples/net/sockets/big_http_download/src/big_http_download.c:44]: Will be changed to int arithmetics.

**Comment by Nukersson** [comment]
Form @pfalcon in #26719: 
Thanks for the patches. They definitely make this sample more complicated, but I'm not sure they improve it. Please keep in mind that this a sample, not a wget implementation or something. Its purpose is to show that it's possible to write simple, POSIX-compatible applications for Zephyr. Specifically:

This sample should be simple for user to study (the less code the better, or at least "more code" is not an improvement).
This sample should be simple, and small, and thus run on as many device as possible (usage of floating-point arithmetic is specifically avoided).
This sample should be buildable on a standard POSIX system (e.g. Linux) using make -f Makefile.posix, and this should be achieved with as few #ifdef's in the source code as possible (so, adding more isn't a good direction).
I think that this patch in its current shape deteriorates the sample on all the points above.

Looking just at the functionality, I agree that showing download speed would be useful. As I mentioned elsewhere, the reason I didn't add it myself is, well, because the figures which come out aren't exactly to brag about :-I. And then, there're multiple download speeds - e.g. averaged over all iterations, averaged over the current iteration, instantaneous (i.e. averaged over small period of time, 3-5s). IMHO, the most useful for Zephyr testing would be the last.

File: samples/net/sockets/big_http_download/prj.conf [hunk 2]
```diff
@@ -3,6 +3,7 @@ CONFIG_NEWLIB_LIBC=y
 CONFIG_MBEDTLS=y
 CONFIG_MBEDTLS_ENABLE_HEAP=y
 CONFIG_MAIN_STACK_SIZE=2176
+CONFIG_SAMPLE_BIG_HTTP_DL_NUM_ITER=3
```
Comment (Nukersson) [inline samples/net/sockets/big_http_download/prj.conf:6]: Form @pfalcon in #26719:
Please don't change existing settings without good motivation. Such changes and motivation should go into a separate PR (on its own). Current usecase for big_http_download is that a user can start it and use for assessing IP stack overall stability, e.g. wait and watch until it downloads 1GB total data (then one kinda assume that "IP stack in Zephyr works").

File: samples/net/sockets/big_http_download/src/big_http_download.c [hunk 3]
```diff
@@ -4,6 +4,10 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include <logging/log.h>
```
Comment (Nukersson) [inline samples/net/sockets/big_http_download/src/big_http_download.c:7]: From @pfalcon in #26719:
big_http_download is a POSIX-compatible sample (i.e. written with purely POSIX API, and portable to any POSIX system). Please run make -f Makefile.posix in the sample dir to ensure it remains so. (This change likely breaks that.)

File: samples/net/sockets/big_http_download/README.rst [hunk 4]
```diff
@@ -36,6 +36,11 @@ Build the Zephyr version of the application like this:
    :goals: build
    :compact:
 
+.. warning::
+
+   On some boards the stack size CONFIG_MAIN_STACK_SIZE must be increased!
+   The appropriate value may be found by step-by-step increase.
+
```
Comment (Nukersson) [inline samples/net/sockets/big_http_download/README.rst:43]: From @pfalcon in #26719:

This is not an ok warning to put in the docs. If the stack size needs to be increased, let's increase it to safe enough value. (The fact that Zephyr stack size requirements in Zephyr is constantly grows is a known matter, I raised concern of it, but nobody seems to pay attention.)

File: samples/net/sockets/big_http_download/prj.conf (hunk 2 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/prj.conf:6]: @pfalcon The value `CONFIG_SAMPLE_BIG_HTTP_DL_NUM_ITER` could not be find in any KConfig in the original implementation of the application! Do you mean, that this was expected?

File: samples/net/sockets/big_http_download/README.rst (hunk 4 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/README.rst:43]: @pfalcon Ok. I will increase the stack to next possible bigger value.

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 3 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/src/big_http_download.c:7]: @pfalcon I've tested this example for POSIX on latest master and it is not compiles:

Compile command: `make -f Makefile.posix`

Error:

`cc src/big_http_download.c -lmbedcrypto -o big_http_download
src/big_http_download.c:12:10: fatal error: 'mbedtls/md.h' file not found
#include "mbedtls/md.h"
         ^~~~~~~~~~~~~~
1 error generated.`

Thus currently not my approach breaks the code!

Comment (Nukersson) [reply samples/net/sockets/big_http_download/src/big_http_download.c:7]: To avoid not supported logging on POSIX following will be done:

`#if defined(__ZEPHYR__) && !defined(CONFIG_POSIX_API)

#include <logging/log.h>
#define LOG_LEVEL LOG_LEVEL_DBG
LOG_MODULE_REGISTER(big_http_download);

#endif`

**Comment by Nukersson** [comment]
Hey @pfalcon thank you for your answer.

> Form @pfalcon in #26719:
> Thanks for the patches. They definitely make this sample more complicated, but I'm not sure they improve it. Please keep in mind that this a sample, not a wget implementation or something. Its purpose is to show that it's possible to write simple, POSIX-compatible applications for Zephyr. Specifically:
> 
> This sample should be simple for user to study (the less code the better, or at least "more code" is not an improvement).

- Why showing KiB and MiB makes the sample complicated? There are only three additional defines and some couple of strings needed.

> This sample should be simple, and small, and thus run on as many device as possible (usage of floating-point arithmetic is specifically avoided).

- I agree with it. Floating point arithmetic were removed.
- The sample is also a way far from being `wget` ;)
> This sample should be buildable on a standard POSIX system (e.g. Linux) using make -f Makefile.posix, and this should be achieved with as few #ifdef's in the source code as possible (so, adding more isn't a good direction).

- Currently POSIX is not possible to compile due to absence of the header `mbedtls/md.h`. Please try to compile for POSIX.
- Why so scary about `#ifdef's`?

> I think that this patch in its current shape deteriorates the sample on all the points above.

- I do not agree: see my explanations above and the less amount of code needed for my improvements.
> 
> Looking just at the functionality, I agree that showing download speed would be useful. As I mentioned elsewhere, the reason I didn't add it myself is, well, because the figures which come out aren't exactly to brag about :-I. And then, there're multiple download speeds - e.g. averaged over all iterations, averaged over the current iteration, instantaneous (i.e. averaged over small period of time, 3-5s). IMHO, the most useful for Zephyr testing would be the last.

Ok. Then I can add the averaged speed.

If you completely disagree with my improvements, please write it here and I will not invest my time in this upgrade anymore. Otherway please test the improved sample and give me a feedback/review on needed changes.

File: samples/net/sockets/big_http_download/prj.conf [hunk 5]
```diff
@@ -2,7 +2,8 @@
 CONFIG_NEWLIB_LIBC=y
```
Comment (jukkar) [inline samples/net/sockets/big_http_download/prj.conf:2]: Typo in the commit msg `s/sickets/sockets/`

File: samples/net/sockets/big_http_download/prj.conf [hunk 6]
```diff
@@ -2,7 +2,8 @@
...
 CONFIG_MBEDTLS=y
 CONFIG_MBEDTLS_ENABLE_HEAP=y
-CONFIG_MAIN_STACK_SIZE=2176
+CONFIG_MAIN_STACK_SIZE=2180
```
Comment (jukkar) [inline samples/net/sockets/big_http_download/prj.conf:5]: We are only increasing the stack size for 4 bytes. Is this really enough here, I mean that if something changes in the system (for example not related to this sample), then we might run out of stack again. I suggest that the value is increased more, like to 2256 for example (I just picked some higher random value that is dividable by 8)

File: samples/net/sockets/big_http_download/prj.conf (hunk 2 above)
Comment (jukkar) [reply samples/net/sockets/big_http_download/prj.conf:6]: > The value `CONFIG_SAMPLE_BIG_HTTP_DL_NUM_ITER` could not be find in any KConfig in the original implementation of the application!

Hmm, this config option is found in `samples/net/sockets/big_http_download/Kconfig`.
Not sure what is the issue here, difficult to follow the discussion as you closed the original PR and created a new one. In the future, please use the original PR when submitting changes (unless the PR is too convoluted or messy), that way the discussion is easier to follow.

File: samples/net/sockets/big_http_download/prj.conf (hunk 6 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/prj.conf]: I've done a mistake in testing previously. Tested it now with `CONFIG_MAIN_STACK_SIZE=2176` and it works fine on my nucleo-h745zi-q board. Thus this change was reverted.

File: samples/net/sockets/big_http_download/prj.conf (hunk 5 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/prj.conf]: fixed

File: samples/net/sockets/big_http_download/prj.conf (hunk 2 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/prj.conf:6]: Sorry for closing previous PR. It was not clear for me how to change PR name and branch's name. Next time I will be better :)

You'r right about `samples/net/sockets/big_http_download/Kconfig` and  `CONFIG_SAMPLE_BIG_HTTP_DL_NUM_ITER `. Thus the commit `samples: net: sockets: big_http_download: preparing upgrade` with corresponding change will be dropped within next push.

File: samples/net/sockets/big_http_download/src/big_http_download.c [hunk 7]
```diff
@@ -4,6 +4,14 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#if defined(__ZEPHYR__) && !defined(CONFIG_POSIX_API)
+
+#include <logging/log.h>
+#define LOG_LEVEL LOG_LEVEL_DBG
+LOG_MODULE_REGISTER(big_http_download);
```
Comment (pfalcon) [inline samples/net/sockets/big_http_download/src/big_http_download.c:11]: It's still unclear why this change is needed. It defines logger named "big_http_download". But it's not used by anything. This sample does all its output using printf().

File: samples/net/sockets/big_http_download/src/big_http_download.c [hunk 8]
```diff
@@ -33,6 +41,11 @@
 
 #endif
 
+#define KiB	*1024u
+#define MiB	*1024u KiB
+#define Bytes2KiB(Bytes)	(Bytes/(1u KiB))
```
Comment (pfalcon) [inline samples/net/sockets/big_http_download/src/big_http_download.c:46]: Please use all lower-case for parameter name "bytes".

And spaces around "/" operator.

File: samples/net/sockets/big_http_download/src/big_http_download.c [hunk 9]
```diff
@@ -195,7 +208,8 @@ void download(struct addrinfo *ai, bool is_tls)
 		mbedtls_md_update(&hash_ctx, response, len);
 
 		cur_bytes += len;
-		printf("%u bytes\r", cur_bytes);
+		printf("Load progress: %u Bytes; %u KiB; %u MiB\r",
```
Comment (pfalcon) [inline samples/net/sockets/big_http_download/src/big_http_download.c:211]: "Load progress" sounds somewhat confusing, unclear which "load" is meant. What happens here is *download*. So, please either make it "%u Bytes; %u KiB; %u MiB\r" to save on the output size (note that on high networking speed, UART output may affect the throughput). Or otherwise, make it unambiguous with "Download progress: %u Bytes; %u KiB; %u MiB\r".

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 8 above)
Comment (pfalcon) [reply samples/net/sockets/big_http_download/src/big_http_download.c:46]: And I'm not sure about macro name, I'd it should be either BYTES2KIB(), or just bytes2kib(), but we don't use mixed case for naming such things. (Well, perhaps it can be bytes2KiB()).

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 7 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/src/big_http_download.c:11]: Should it be removed then?

**Comment by pfalcon** [comment]
> samples: net: sockets: big_http_download: add progress metrics 

I don't think this commit message correctly describes the change. The matter, the was progress metrics already. You added more to them, or extended them, then please write just that: "samples: net: sockets: big_http_download: Extend progress metrics".

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 8 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/src/big_http_download.c:46]: I like last one :) Will change within next push.

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 9 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/src/big_http_download.c:211]: "Download progress" makes more sense.  You're right.
I absolutely agree, this printf will slow down the speed.
I thought to bring this printf into separted tread. What do you think about?

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 7 above)
Comment (pfalcon) [reply samples/net/sockets/big_http_download/src/big_http_download.c:11]: > Should it be removed then?

Unless you can explain why it's needed (i.e. give instructions how to reproduce some issue, which addition of this logger fixes), then yes, I don't see how it's useful.

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 9 above)
Comment (pfalcon) [reply samples/net/sockets/big_http_download/src/big_http_download.c:211]: > I absolutely agree, this printf will slow down the speed.

Then you understand why the original message was just "%u bytes\r" and nothing else ;-).

> I thought to bring this printf into separted tread. What do you think about?

Please don't. As I already mentioned, this should be a **simple** sample, it should not be an industrial-grade networking throughput testing tool. If you're interested in having better means of speed testing, that should be a separate application. And we actually have such, it's https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/net/zperf . I'd suggest you actually pay more attention to it instead - it needs more people to use it, and certainly some improvements too.

**Comment by Nukersson** [comment]
> > samples: net: sockets: big_http_download: add progress metrics 
> 
> I don't think this commit message correctly describes the change. The matter, the was progress metrics already. You added more to them, or extended them, then please write just that: "samples: net: sockets: big_http_download: Extend progress metrics".
> 

Ok. Will do it.

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 7 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/src/big_http_download.c:11]: I thought Zephyrs logging API supports also POSIX. It looks like I was wrong. The logging module will be removed in next commit for sake of simplicity.

File: samples/net/sockets/big_http_download/src/big_http_download.c (hunk 9 above)
Comment (Nukersson) [reply samples/net/sockets/big_http_download/src/big_http_download.c:211]: Ok. Understand it. Thank you.

**Comment by Nukersson** [comment]
Dear all. With this push I kept the changes as small as possible and also improved all suggested parts. For me the upgrade looks very good now. I hope you enjoy the possible last review :D

**Review by pfalcon — APPROVED** [review]
LGTM. Thanks for all the updates.

**Review by jukkar — APPROVED** [review]
